crates/kitsune_p2p/dht/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+//! the network, and a constantly increasing height as time marches forward.
+//!
+//! So:
+//! - An Op can be thought of a point somewhere on the surface of this cylinder,
+//! - An Agent can be thought of as a line perpendicular to the base, extending
+//!   across the entire length of the cylinder.
+//! - An Arcs can be thought of as a rectangular "stripe" made by sweeping the
+//!   Agent line across the distance specified by the length of the Arc.
+//! - A *Region* is an arbitrary rectangular area of spacetime.
+//!
+//! ## Quantization
+//!
+//! One key feature of this DHT space is that it is *quantized* in both dimensions.
+//! You can think of this as a grid which overlays spacetime, which has the effect
+//! of grouping Agents and Ops into buckets specified by the grid cells. Another
+//! way to think of this is we only refer to items by which grid cell they appear in,
+//! not by their absolute coordinates.
+//!
+//! ### Topology and Quantum Coordinates
+//!
+//! The quantization is specified by a parameter called the [`Topology`]. This
+//! construct defines how the quantized grid is overlaid onto spacetime, and specifies
+//! how to transform between "absolute" coordinates (Locations and Timestamps) and
+//! "quantum" coordinates (grid cells).
+//!
+//! Each Kitsune Space is free to choose its own Topology (and all nodes in that space
+//! must agree to use the same Topology). However, in practice we currently use the same
+//! topology for all spaces:
+//! - the standard space quantum size is 2^12
+//! - the standard time quantum size is 5 minutes
+//!
+//! Additionally, we include a "time origin" component in the topology, which describes
+//! a shift of the origin of the time dimension of the quantized grid. In other words,
+//! the time origin specifies what Timestamp corresponds to the origin coordinate of
+//! the grid's time dimension.
+//!
+//! For the purpose of terminology, we refer to these rectangular grid cells as the
+//! spacetime quanta. The sides of these rectangles are the space quanta and time quanta.
+//!
+//! ### Segments (Chunks) and Exponential Coordinates (TODO land on a name for this?)
+//!
+//! For the purposes of our gossip algorithm, we specify Arcs in terms of "segments"
+//! or "chunks". A Segment is a set of `2^p` contiguous quanta, where `p` is a whole number.
+//! So, segments are always a power-of-two multiple of the quantum size. Additionally, the offset
+//! from the origin must also be a power of two. So, the set of all possible segments
+//! implies a collection of uniform grids layered on top of the quantum grid, each one twice
+//! as coarse as the one below it.
+//!
+//! Thus, we can refer to the successive levels of coarseness of quantization by a single
+//! whole number, referred to as the "power", since it takes the form of `2^p`.
+//! A "power" of 0 refers to the quantum grid itself. A power of 1 is twice as coarse, and so on.
```

Comment:
I don't understand this part. The cylinder is divided into discs the height of the time quantum. Each disc is divided into space quantum number of cells. The discs are the layers. What does "quantum grid" refer to? And why is the number of cells growing with increasing time?

Reply:
By "quantum grid" I mean the resolution at the quantum level, which can still be thought of as a grid which is overlaid on top of the absolute coordinates. The quantum grid is the finest possible resolution as far as kitsune is concerned, but it is actually very coarse in terms of absolute coordinates.

Reply:
I'm very happy to rewrite parts of this to use a less geometrical interpretation, I have been thinking on this on my own for months so I probably have my own peculiar way of thinking about it that may not map well onto the general psyche.

Reply:
> Each disc is divided into space quantum number of cells

Actually, no, the space quantum is the size of each cell, not the number. Larger quantum size means fewer cells.

> And why is the number of cells growing with increasing time?

Simply because as time goes on into the future, more "rows" of cells are needed to capture the newer timestamps which didn't exist previously.

Reply:
After your explanations and a second read I get it. The aspect that I hadn't grokked is that there are absolute values (Location and Time) **and** there's a grid imposed on those values, quantizing those values or putting them in buckets.

It's not peculiar I think, it's just an exercise to understand this concept without images. It makes sense to me now.

> Simply because as time goes on into the future, more "rows" of cells are needed to capture the newer timestamps which didn't exist previously.

Another thing I hadn't considered. This implies that gossip is always complete back in history so to speak, not only considering the current row of cells. Therefore the number of rows is constantly growing. Thinking about it, it has to be like that. Otherwise changes in the arcs would lead to incomplete DHTs.